File: src/game/__tests__/factions.test.ts
import { describe, it, expect } from "vitest";
import { factionAdvantage, lineupAura } from "../factions";

describe("factions", () => {
  it("Vanguard beats Horde", () => {
    const adv = factionAdvantage("vanguard", "horde");
    expect(adv.damage).toBeGreaterThan(0);
  });

  it("Horde suffers vs Vanguard", () => {
    const adv = factionAdvantage("horde", "vanguard");
    expect(adv.damage).toBeLessThan(0);
  });

  it("5-of-a-kind mono-faction gives an aura", () => {
    const a = lineupAura(["abyss", "abyss", "abyss", "abyss", "abyss"]);
    expect(a.faction).toBe("abyss");
    expect(a.bonus.attack).toBeGreaterThan(0);
  });

  it("Dawnfall Pact unlocks with 2 radiance + 2 abyss + 1 other", () => {
    const a = lineupAura(["radiance", "radiance", "abyss", "abyss", "vanguard"]);
    expect(a.faction).toBe("dawnfall");
  });
});

File: android/app/src/test/java/com/herobrawl/game/engine/FactionsTest.kt
package com.herobrawl.game.engine

import com.herobrawl.game.data.Factions
import com.herobrawl.game.model.FactionId
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class FactionsTest {
    @Test
    fun `vanguard beats horde`() {
        assertTrue(Factions.advantage(FactionId.VANGUARD, FactionId.HORDE) > 0)
    }

    @Test
    fun `horde suffers vs vanguard`() {
        assertTrue(Factions.advantage(FactionId.HORDE, FactionId.VANGUARD) < 0)
    }

    @Test
    fun `five of a kind grants mono aura`() {
        val a = Factions.lineupAura(List(5) { FactionId.ABYSS })
        assertTrue(a.attack > 0)
    }

    @Test
    fun `dawnfall pact activates`() {
        val a = Factions.lineupAura(
            listOf(FactionId.RADIANCE, FactionId.RADIANCE, FactionId.ABYSS, FactionId.ABYSS, FactionId.VANGUARD)
        )
        assertEquals(0.18, a.attack, 0.001)
    }
}
